libdrgn/arch_ppc64.c

```diff
@@ -267,6 +267,194 @@ apply_elf_reloc_ppc64(const struct drgn_relocating_section *relocating,
+	drgn_object_init(&cur_cpu_spec, prog);
+	drgn_object_init(&mmu_features_obj, prog);
+
+	if (page_shift == 16) {
+		it->pt_levels = pt_levels_radix_64k;
+	} else if (page_shift == 12) {
+		it->pt_levels = pt_levels_radix_4k;
+	} else {
+		err = drgn_error_create(DRGN_ERROR_OTHER,
+					 "Unknown page size.");
+		goto out;
+	}
+
+	/*
+	 * Only BOOK3S CPU family is supported.
+	 * If interrupt_base_book3e is missing then it is BOOK3S.
+	 */
+	err = drgn_program_find_object(prog, "interrupt_base_book3e", NULL,
+				       DRGN_FIND_OBJECT_ANY, &book3s);
+	if (!err) {
+		err = drgn_error_create(DRGN_ERROR_OTHER,
+					"Virtual address translation is not available for BOOK3E CPU family.");
+		goto out;
+	}
+
+	/* Identify the MMU type. */
+	err = drgn_program_find_object(prog, "cur_cpu_spec", NULL,
+				       DRGN_FIND_OBJECT_ANY, &cur_cpu_spec);
+	if (err)
+		goto out;
+	err = drgn_object_member_dereference(&mmu_features_obj, &cur_cpu_spec,
+					     "mmu_features");
+	if (err)
+		goto out;
+
+	err = drgn_object_read_unsigned(&mmu_features_obj, &mmu_features);
+
+	if (err)
+		goto out;
+
+	if (!(mmu_features & 0x40)) {
+		err = drgn_error_create(DRGN_ERROR_OTHER,
+					"Virtual address translation is only supported for Radix MMU.");
+		goto out;
+	}
+
+	*ret = &it->it;
+
+out:
+	drgn_object_deinit(&book3s);
+	drgn_object_deinit(&cur_cpu_spec);
+	drgn_object_deinit(&mmu_features_obj);
+	if (err)
+		free(it);
+	return err;
+}
+
+static void
+linux_kernel_pgtable_iterator_destroy_ppc64(struct pgtable_iterator *_it)
+{
+	struct pgtable_iterator_ppc64 *it =
+		container_of(_it, struct pgtable_iterator_ppc64, it);
+	free(it);
+}
+
+static void
+linux_kernel_pgtable_iterator_init_ppc64(struct drgn_program *prog,
+					       struct pgtable_iterator *_it)
+{
+	return;
+}
+
+static struct drgn_error *
+linux_kernel_pgtable_iterator_next_ppc64(struct drgn_program *prog,
+					 struct pgtable_iterator *_it,
+					 uint64_t *virt_addr_ret,
+					 uint64_t *phys_addr_ret)
+{
+	uint16_t level;
+	uint64_t virt_addr;
+	struct drgn_error *err;
+	static const uint16_t levels = 4;
+	struct pgtable_iterator_ppc64 *it;
+	static const uint64_t pt_mask = 0xc0000000000000ffUL;
+
+	/* Page table entry status bits since Linux kernel commits 849f86a630e9
+	 * ("powerpc/mm/book3s-64: Move _PAGE_PRESENT to the most significant bit")
+	 * and 84c957560a7a ("powerpc/mm/book3s-64: Move _PAGE_PTE to 2nd most
+	 * significant bit") (in v4.6).
+	 */
+	static const uint64_t page_pte = UINT64_C(1) << 62;
+	static const uint64_t page_present = UINT64_C(1) << 63;
+
+	it = container_of(_it, struct pgtable_iterator_ppc64, it);
+	level = levels;
+	virt_addr = it->it.virt_addr;
+
+	for (;;level--) {
+		uint64_t index;
+		uint64_t table;
+		uint64_t entry;
+		bool table_physical;
+
+		if (level == levels) {
+			table = it->it.pgtable;
+			table_physical = false;
+		} else {
+			/* PAGE_PTE bit represents huge page */
+			if (!entry || (entry & page_pte) || level == 0) {
```

Comment:
Re: `!entry`, it looks like the book3s kernel page walking code (I usually reference `follow_page_mask` and the `pte_*`, `pmd_*`, etc. helpers it calls) checks the `_PAGE_PRESENT` but for the inner levels instead of comparing to zero.

Reply:
Agree. Changing the condition to. 

if (!(entry & page_present) || (entry & page_pte) || level == 0)